File: Video-Editors/next-editor-version/server/internal/editor/session.go
package editor

// The session log — a record of real editing, kept as you work.
//
// The test writes its own log; this is the other half: what actually happened
// to your files, in the same shape. One file per DAY, appended, with a header
// each time the server starts.
//
// Only calls that CHANGE something are logged — plus the ways IN, because a
// line saying what you were working on is what makes the rest readable.
//
// It has already earned its place twice: it showed a Join that never reached
// the server at all, and four pastes that landed on one track and were refused
// on the other.

import (
	"fmt"
	"net/url"
	"os"
	"path/filepath"
	"strings"
	"sync"
	"time"
)

type action struct {
	label string
	keys  []string // which payload keys are worth showing
}

var actions = map[string]action{
	"/api/frames/dup":      {"+ Frame", []string{"at", "count", "side"}},
	"/api/frames/del":      {"- Frame", []string{"at", "count", "side"}},
	"/api/frames/dup-span": {"+ Zone", []string{"a", "b"}},
	"/api/frames/del-span": {"- Zone", []string{"a", "b"}},
	"/api/frames/restore":  {"Undo", nil},
	"/api/frames/paste":    {"Paste frame", []string{"from", "at"}},
	"/api/mark":            {"Mark", []string{"frame", "on"}},
	"/api/clear-marks":     {"Unmark all", nil},
	"/api/save":            {"Save scene", nil},
	"/api/cut":             {"Cut scene", nil},
	"/api/clear-edits":     {"Discard edits", nil},
	"/api/reset-editor":    {"Reset editor", nil},
	"/api/join":            {"Join", []string{"ns", "label", "tracks"}},
	"/api/split":           {"Split", []string{"n", "at", "labels", "tracks"}},
	"/api/line":            {"Edit line", []string{"n"}},
	"/api/renumber-clear":  {"Lift lock", nil},
	"/api/handoff":         {"Hand off", []string{"version", "names"}},
	"/api/archive":         {"Archive", []string{"folder"}},
	"/api/open":            {"Open clip", []string{"path"}},
	"/api/open-pair":       {"Open layered", []string{"base"}},
	"/api/open-seq":        {"Open timeline", []string{"root", "ns"}},
	"/api/open-pair-go":    {"Open layered", []string{"base"}},
	"/api/open-seq-go":     {"Open timeline", []string{"root", "ns"}},
}

// resultKeys — what is worth showing back, in the order it reads best.
var resultKeys = []string{"nb_frames", "count", "version", "duration_s", "joined",
	"split", "label", "labels", "line", "renamed", "url", "slug"}

type SessionLog struct {
	path string
	off  bool
	mu   sync.Mutex
}

func NewSessionLog(root string, off bool) *SessionLog {
	return &SessionLog{
		path: filepath.Join(root, "logs", fmt.Sprintf("editor_%s.log", time.Now().Format("20060102"))),
		off:  off,
	}
}

func (s *SessionLog) Path() string {
	if s.off {
		return "off"
	}
	return s.path
}

func (s *SessionLog) Start(port int, root string) {
	if s.off {
		return
	}
	_ = os.MkdirAll(filepath.Dir(s.path), 0o755)
	ver := SAELabel("Segment and Avatar Editor", "segment_avatar_editor")
	s.write(fmt.Sprintf("\nEditor Session:  %s\nPlayer:          %s\nServer:          http://localhost:%d  (Go)\n\n",
		time.Now().Format("2006-01-02T15:04:05"), ver, port))
}

func (s *SessionLog) write(text string) {
	s.mu.Lock()
	defer s.mu.Unlock()
	f, err := os.OpenFile(s.path, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0o644)
	if err != nil {
		return // a log that can break the editor is worse than no log
	}
	defer f.Close()
	_, _ = f.WriteString(text)
}

func (s *SessionLog) LogGET(path string, q url.Values, rep *reply) {
	if s == nil || s.off {
		return
	}
	if _, ok := actions[path]; !ok {
		return
	}
	p := Payload{}
	for k := range q {
		p[k] = q.Get(k)
	}
	s.line(path, p, rep)
}

func (s *SessionLog) LogPOST(path string, p Payload, rep *reply) {
	if s == nil || s.off {
		return
	}
	s.line(path, p, rep)
}

func (s *SessionLog) line(path string, p Payload, rep *reply) {
	act, ok := actions[path]
	if !ok {
		return
	}
	var args []string
	for _, k := range act.keys {
		if v, present := p[k]; present && v != nil {
			args = append(args, fmt.Sprintf("%s=%v", k, plain(v)))
		}
	}
	// What it acted on: a cache slug for frame work, a store for the rest.
	who := p.str("slug")
	if who == "" {
		who = p.str("root")
	}
	argStr := strings.Join(args, " ")
	if path == "/api/open" {
		src := p.str("path")
		who = filepath.Base(filepath.Dir(src))
		argStr = filepath.Base(src)
	}

	tail := ""
	if rep.status != 200 || rep.body["error"] != nil {
		e := rep.body["error"]
		if e == nil {
			e = rep.status
		}
		tail = fmt.Sprintf("REFUSED: %v", e)
	} else {
		var parts []string
		for _, k := range resultKeys {
			if v, ok := rep.body[k]; ok {
				parts = append(parts, fmt.Sprintf("%s=%v", k, plain(v)))
			}
		}
		tail = strings.Join(parts, "  ")
	}
	s.write(strings.TrimRight(fmt.Sprintf("%s  %-14s %-26s %s  %s",
		time.Now().Format("15:04:05"), act.label, who, argStr, tail), " ") + "\n")
}

// plain renders a value the way the Python log did — a float that is really an
// integer prints without its `.0`, so `at=440` does not read as `at=440.0`.
func plain(v any) any {
	switch x := v.(type) {
	case float64:
		if x == float64(int64(x)) {
			return int64(x)
		}
	case []any:
		out := make([]any, len(x))
		for i, e := range x {
			out[i] = plain(e)
		}
		return out
	}
	return v
}

File: Video-Editors/editor_base/session.py
"""
One line per action, in every editor's own log.

⚠ THIS WAS BORROWED, NOT SHARED. Frame Blender and Avatar Editor each did
`import serve as main_serve` and then wrote into that module's globals —
`main_serve.ACTIONS = ...`, `main_serve.SESSION_LOG = ...` — to make ITS logger
write THEIR file. So two editors were configured by reaching into a third
editor's module state, and the log you got depended on which import ran last.

Here each editor configures its own logger and nothing reaches into anything:

    from editor_base import session
    session.configure(os.path.join(LOGS, "frame_blender_20260921.log"),
                      ACTIONS, off=args.no_session_log)
    session.log(path, payload, result, status)

`actions` maps a route to (label, keys-worth-showing). A route that is not in
it is not logged — that is how the noise stays out: every GET of a frame image
would otherwise bury the lines that say what someone did.
"""
import os
import time

# Set by configure(). Module state on purpose: one process serves one editor.
_LOG = None
_ACTIONS = {}
_OFF = False

# What a result is worth saying back. A handler returns plenty; these are the
# numbers a person reads a log for.
RESULT_KEYS = ("nb_frames", "current", "span", "actual", "dropped_marks",
               "duration_s", "wrote", "scenes", "error")


def configure(log_path, actions, off=False, result_keys=None):
    """Point this process's logger at its own file and its own route table."""
    global _LOG, _ACTIONS, _OFF, RESULT_KEYS
    _LOG, _ACTIONS, _OFF = log_path, dict(actions or {}), bool(off)
    if result_keys:
        RESULT_KEYS = tuple(result_keys)
    if _LOG and not _OFF:
        os.makedirs(os.path.dirname(_LOG), exist_ok=True)
    return _LOG


def log_path():
    """Where this process is writing, for the line an editor prints at boot."""
    return _LOG or "(not configured)"


def header(lines):
    """Write the few lines that open a session — which editor, which port."""
    if not _LOG or _OFF:
        return
    try:
        with open(_LOG, "a") as fh:
            for line in lines:
                fh.write(line.rstrip() + "\n")
            fh.write("\n")
    except OSError:
        pass


def log(path, payload, result, status):
    """
    One line per action. NEVER RAISES: a log that can break the editor is worse
    than no log, and this runs inside a request handler.
    """
    entry = _ACTIONS.get(path)
    if entry is None or _OFF or not _LOG:
        return
    label, keys = entry
    try:
        args = " ".join(f"{k}={payload.get(k)}" for k in keys if payload.get(k) is not None)
        # what it acted on: a cache slug for frame work, a store for the rest
        who = payload.get("slug") or payload.get("root") or ""
        res = result if isinstance(result, dict) else {}
        if status != 200 or res.get("error"):
            tail = f"REFUSED: {res.get('error', status)}"
        else:
            tail = "  ".join(f"{k}={res[k]}" for k in RESULT_KEYS if k in res)
        with open(_LOG, "a") as fh:
            fh.write(f"{time.strftime('%H:%M:%S')}  {label:<14} {who:<26} "
                     f"{args}  {tail}".rstrip() + "\n")
    except Exception:
        pass
